### ISSUE TITLE
Magic cookbook is incompatible with `logstash_config`

### ISSUE THREAD

**Issue by EdHurtig** [evertrue/et_elk-cookbook#11]
the recent introduction of the magic cookbook to the infrastructure creates a conflict with the `logstash_config` LWRP from the `logstash` cookbook and the `logstash_config` helper method from the majic cookbook

I have no idea how to resolve this other than creating our own version of the logstash or magic cookbooks

/cc @eherot

**Comment by eherot** [comment]
Can you be more specific about the exact nature of this conflict? E.g. does it conflict with a sub-dependency cookbook?

**Comment by EdHurtig** [comment]
https://github.com/sczizzo/magic-cookbook/blob/master/libraries/configuration.rb#L140-L201

When the magic cookbook is included as part of the convergence then the logstash_config method linked to above conflicts with the logstash_config resource

```
================================================================================
  Recipe Compile Error in /var/chef/cache/cookbooks/et_elk/recipes/server.rb
  ================================================================================

  NoMethodError
  -------------
  undefined method `each' for nil:NilClass

  Cookbook Trace:
  ---------------
    /var/chef/cache/cookbooks/magic/libraries/configuration.rb:187:in `logstash_config'
    /var/chef/cache/cookbooks/elk_logstash/recipes/default.rb:24:in `from_file'
    /var/chef/cache/cookbooks/et_elk/recipes/server.rb:9:in `block in from_file'
    /var/chef/cache/cookbooks/et_elk/recipes/server.rb:9:in `each'
    /var/chef/cache/cookbooks/et_elk/recipes/server.rb:9:in `from_file'

  Relevant File Content:
  ----------------------
  /var/chef/cache/cookbooks/magic/libraries/configuration.rb:

  180:            <%= name %> {<% config.each do |k, v| %>
  181:              <%= k %> => <%= quote_logstash v %><% end %>
  182:            }
  183:        <% end %><% if type.is_a?(String) %>  }
  184:        <% end %><% end %>
  185:        }
  186:      $.strip.gsub(/^      /, '')
  187>>     ERB.new(tpl).result(binding).gsub(/(\n\s*)+?\n/, "\n")
  188:    end
  189:  end
  190:
  191:  class Chef
  192:    class Recipe
  193:      include Configuration
  194:    end
  195:  end
  196:
```

**Comment by EdHurtig** [comment]
Where is the magic cookbook being used?

**Comment by EdHurtig** [comment]
nvm, found it in et_base... right where I expected it to be

**Comment by eherot** [comment]
Wow, that's annoying. Probably easier to change our own namespace than to ask them to change theirs though.

**Comment by eherot** [comment]
Unless of course we can actually make use of magic's logstash_config method...

**Comment by EdHurtig** [comment]
is there a way to namespace an LWRP... something on the order of `run_context.resources.logstash_config`?

**Comment by EdHurtig** [comment]
and I would prefer to stick with the `logstash` cookbook's way of doing things... I think it would be much cleaner than 

```
file 'some/path/that/we/need/to/integrate/with' do 
    content logstash_config(node['elk_logstash']['logstash_configs']['sidekiq'])
end
```

and then go through attr hell with Chef. I like creating templates where I have full freedom to comment, do complex `if` logic, and generally have good readability of the configuration.... you can basically understand the entire configuration of logstash by just reading the `templates/default/` folder of this cookbook which I think is really useful given that logstash configs can get pretty complicated when you start tracking lots of services

**Comment by eherot** [comment]
On the other hand...
- Overloading templates makes settings difficult to override
- This requires us to maintain our own Logstash config generator when another well-maintained project is already maintaining one.
- The example you just cited isn't all that messy and we use similar form in many other places (especially when generating yaml_files). Actually, most things that use `magic` use that syntax.

**Comment by EdHurtig** [comment]
However, Logstash configs are more than just YAML files, they are essentially a somewhat simplified programming language for defining the behavior of how logstash should proccess events.  Although there is already a generator for it in the magic cookbook, it is primative and limited in what it can do.

It is easy to start seeing where the limitations can come in to play in their own example:

```ruby
logstash_config({
  'input' => {
    'test' => {
      'file' => {
        'path' => '/var/log/test.log'
      }
    }
  },
  'filter' => {
    'test' => {
      'seq' => {}
    }
  },
  'output' => {
    'test' => {
      'stdout' => {
        'codec' => 'rubydebug'
      }
    }
  }
})
```

Generates:

```
input {
  file {
    path => "/var/log/test.log"
    type => "test"
  }
}
filter {
  if [type] == "test" {
    seq {
    }
  }
}
output {
  if [type] == "test" {
    stdout {
      codec => "rubydebug"
    }
  }
}
```

Notice how it assumes that I always want a exactly 1 condition around all my inputs, outputs, and filters... what happens when I want no conditions, or 2 conditions... what happens when I want to nest conditions... or say things like `if [app] == "dna" and [type] = "rails_app"`

I strongly believe that it would be a serious regression and add complications to our logstash configuration if we start using `logstash_config` from the `magic` cookbook.  

Also w/r/t overriding templates, the thing is `et_elk` IS the top level wrapper for Logstash Server Configs, we should not have to wrap `et_elk`, or change the Logstasg configs from other cookbooks.  In fact, because the logstash server configs are placed on the central servers, it is impossible to change the configuration of logstash via other chef cookbooks like `et_rails_app` because `et_rails_app` is not assigned to the `env-logs-...` servers.

**Comment by EdHurtig** [comment]
Also as much as I love the idea just using something as simple as `yaml_dump node['...']['...']` in chef cookbooks, I have started to become very sour to this approach of generating configs. The lack of Idempotence with chef attributes is incredibly dangerous for these kinds of things as it is so often overlooked... and in order to specify the config in an idempotent way, we would need to say something

```ruby
set['...']['...'] = {
   big: 'hash',
   that: {
     is: {
       not: 'a chef mash',
       but: 'just a hash' 
    }
   }
}
```

Which means in order to override any settings you would need to do this every time you wanted to add or change something about it. notice the fact that this runs in compile time which brings even more problems...

```ruby
dat_hash = node['...']['...']
dat_hash[:that][:is][:but] = 'annoying'
dat_hash[:this] = { a: 'new', thing: '!!' }
node.set['...']['...'] = dat_hash
```

So long story short... I like the way I have this cookbook setup... and I want this voodoo magic stuff out of my life or at least a way around it like namespacing

**Comment by eherot** [comment]
I don't really want to get into a debate here about the rationale of using attributes to populate config files (I think that's kind of settled science in the Chef community, but also it's ancillary to this issue).

I punted on this by removing `magic` from the `et_base` cookbook and also [suggesting to the magic folks](https://github.com/sczizzo/magic-cookbook/issues/5) that they change the name of their logstash config method.

**Comment by EdHurtig** [comment]
Ok, thanks for taking care of that in et_base. And yeah I agree with using chef atts for config file population, but it just doesn't seem right here... but anyways that is irrelevant for this issue which is simply a name conflict.
